pkg/cmd/roachtest/tests/backup_fixtures.go

```diff
@@ -0,0 +1,319 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func makeBackupFixtureSpecs(override scheduledBackupSpecs) scheduledBackupSpecs {
+	backupSpecs := makeBackupSpecs(override.backupSpecs, defaultBackupFixtureSpecs.backupSpecs)
+	specs := scheduledBackupSpecs{
+		backupSpecs:              backupSpecs,
+		incrementalBackupCrontab: defaultBackupFixtureSpecs.incrementalBackupCrontab,
+	}
+	if override.incrementalBackupCrontab != "" {
+		specs.incrementalBackupCrontab = override.incrementalBackupCrontab
+	}
+	specs.ignoreExistingBackups = override.ignoreExistingBackups
+	return specs
+}
+
+// defaultBackupFixtureSpecs defines the default scheduled backup used to create a fixture.
+var defaultBackupFixtureSpecs = scheduledBackupSpecs{
+	// Runs an incremental backup every 5 minutes.
+	incrementalBackupCrontab: "*/5 * * * *",
+
+	// The default option of false prevents roachtest users from overriding the
+	// latest backup in a collection, which may be used in restore roachtests.
+	ignoreExistingBackups: false,
+
+	backupSpecs: backupSpecs{
+		version:         "v23.1.1",
+		cloud:           spec.AWS,
+		fullBackupDir:   "LATEST",
+		backupsIncluded: 48,
+		workload: tpceRestore{
+			customers: 25000,
+		},
+	},
+}
+
+const scheduleLabel = "schedule_cluster"
+
+type scheduledBackupSpecs struct {
+	backupSpecs
+	// ignoreExistingBackups if set to true, will allow a new backup chain
+	// to get written to an already existing backup collection.
+	ignoreExistingBackups    bool
+	incrementalBackupCrontab string
+}
+
+func (sbs scheduledBackupSpecs) scheduledBackupCmd() string {
+	// This backup schedule will first run a full backup immediately and then the
+	// incremental backups at the given incrementalBackupCrontab cadence until the user cancels the
+	// backup schedules. To ensure that only one full backup chain gets created,
+	// begin the backup schedule at the beginning of the week, as a new full
+	// backup will get created on Sunday at Midnight ;)
+	var ignoreExistingBackupsOpt string
+	if sbs.ignoreExistingBackups {
+		ignoreExistingBackupsOpt = "ignore_existing_backups"
+	}
+	backupCmd := fmt.Sprintf(`BACKUP INTO %s WITH revision_history`, sbs.backupCollection())
+	cmd := fmt.Sprintf(`CREATE SCHEDULE %s FOR %s RECURRING '%s' FULL BACKUP '@weekly' WITH SCHEDULE OPTIONS first_run = 'now', %s`,
```

Comment:
I'm wondering given that we are having a goroutine that monitors the existence of a backup created by a schedule whether it's better to manually invoke a backup for every layer that we want, instead of using a schedule. This allows us to get better error handling if any individual backups fail, and allows for frequencies that are more specific than the crontab, etc.

Reply:
I think the point of this set of roachtests and this specific call is to create a fixture. I could imagine for a generic backup-restore roundtrip test test, we'd want a more flexible api for scheduling backups, as you say.